SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialLatencyTracer.cpp

```diff
@@ -0,0 +1,360 @@
+	}
+}
+
+void USpatialLatencyTracer::OnDequeueMessage(const SpatialGDK::FOutgoingMessage* Message)
+{
+	if (Message->Type == SpatialGDK::EOutgoingMessageType::ComponentUpdate)
+	{
+		const SpatialGDK::FComponentUpdate* ComponentUpdate = static_cast<const SpatialGDK::FComponentUpdate*>(Message);
+		EndLatencyTrace(ComponentUpdate->Trace, TEXT("Sent to Worker SDK"));
+	}
+}
+
+USpatialLatencyTracer* USpatialLatencyTracer::GetTracer(UObject* WorldContextObject)
+{
+	UWorld* World = GEngine->GetWorldFromContextObject(WorldContextObject, EGetWorldErrorMode::ReturnNull);
+	if (World == nullptr)
+	{
+		World = GWorld;
+	}
+
+	if (USpatialGameInstance* GameInstance = World->GetGameInstance<USpatialGameInstance>())
+	{
+		return GameInstance->GetSpatialLatencyTracer();
+	}
+
+	return nullptr;
+}
+
+bool USpatialLatencyTracer::BeginLatencyTrace_Internal(const AActor* Actor, const FString& FunctionName, const FString& TraceDesc)
+{
+	FScopeLock Lock(&Mutex);
+
+	TraceKey Key = CreateNewTraceEntry(Actor, FunctionName);
+	if (Key == InvalidTraceKey)
+	{
+		UE_LOG(LogSpatialLatencyTracing, Warning, TEXT("(%s) : Failed to create Actor/Func trace (%s)"), *WorkerId, *TraceDesc);
+		return false;
+	}
+
+	FString SpanMsg = FormatMessage(TraceDesc);
+	TraceSpan NewTrace = improbable::trace::Span::StartSpan(TCHAR_TO_UTF8(*SpanMsg), nullptr);
+
+	WriteKeyFrameToTrace(&NewTrace, FString::Printf(TEXT("Begin trace : %s"), *FunctionName));
+
+	TraceMap.Add(Key, MoveTemp(NewTrace));
+
+	return true;
+}
+
+bool USpatialLatencyTracer::ContinueLatencyTrace_Internal(const AActor* Actor, const FString& FunctionName, const FString& TraceDesc)
+{
+	FScopeLock Lock(&Mutex);
+
+	TraceSpan* ActiveTrace = GetActiveTrace();
+	if (ActiveTrace == nullptr)
+	{
+		UE_LOG(LogSpatialLatencyTracing, Warning, TEXT("(%s) : No active trace to continue (%s)"), *WorkerId, *TraceDesc);
+		return false;
+	}
+
+	TraceKey Key = CreateNewTraceEntry(Actor, FunctionName);
+	if (Key == InvalidTraceKey)
+	{
+		UE_LOG(LogSpatialLatencyTracing, Warning, TEXT("(%s) : Failed to create Actor/Func trace (%s)"), *WorkerId, *TraceDesc);
+		return false;
+	}
+
+	WriteKeyFrameToTrace(ActiveTrace, TCHAR_TO_UTF8(*TraceDesc));
+	WriteKeyFrameToTrace(ActiveTrace, FString::Printf(TEXT("Continue trace : %s"), *FunctionName));
+
+	TraceMap.Add(Key, MoveTemp(*ActiveTrace));
+	TraceMap.Remove(ActiveTraceKey);
+
+	return true;
+}
+
+bool USpatialLatencyTracer::EndLatencyTrace_Internal()
+{
+	FScopeLock Lock(&Mutex);
+
+	TraceSpan* ActiveTrace = GetActiveTrace();
+	if (ActiveTrace == nullptr)
+	{
+		UE_LOG(LogSpatialLatencyTracing, Warning, TEXT("(%s) : No active trace to end"), *WorkerId);
+		return false;
+	}
+
+	WriteKeyFrameToTrace(ActiveTrace, TEXT("End Trace"));
+	ActiveTrace->End();
+
+	TraceMap.Remove(ActiveTraceKey);
+
+	return true;
+}
+
+TraceKey USpatialLatencyTracer::CreateNewTraceEntry(const AActor* Actor, const FString& FunctionName)
+{
+	static TraceKey NextTraceKey = 1;
```

Comment:
I made it static to ensure when running in the editor we never get conflicting ids. This isn't a hot path, so won't be a performance problem.

Reply:
Why would there be conflicting IDs in a non static member?

Reply:
I don't mention performance because I think we need to optimise this so much as it being static seems like a pessimisation for a single threaded thing.

Reply:
If I haven't properly accounted for accessing the tracer, then there's a chance the editor instance could grab the wrong latency tracer. Perhaps it's better if it falls over quickly, rather than failing obscurely. Will update to non-static member.